_posts/2024-01-16-fido-passkey.md

```diff
@@ -0,0 +1,1127 @@
+The public key is stored on the service provider's server, while the private key remains securely on your device (smartphone, tablet, or a dedicated FIDO security key).
+
+To register as a new user, we must send our public key to the server and have it validated. 
+The server will send us a challenge that we must sign using our private key. 
+If the server can verify the signature using our public key, the registration is complete.
+
+To authenticate a user, we send our username to the server, and the server will answer with a challenge encrypted with our public key.
+On our device, we need to decrypt the challenge with our private key, solve the challenge, and return our response encrypted with our private key.
+After sending the encrypted response, the server is able to validate the response with our public key and check if the login succeeded.
+
+So, in this way, we no longer store the user's password, but only a public key that can be shared openly.
+
+## Create a new Spring Boot application
+
+In our Spring Boot application, we are going to handle the registration and login.
+
+### Add the needed dependencies
+
+```xml
+<dependency>
+    <groupId>org.springframework.boot</groupId>
+    <artifactId>spring-boot-starter-web</artifactId>
+</dependency>
+
+<dependency>
+    <groupId>org.springframework.boot</groupId>
+    <artifactId>spring-boot-starter-data-jpa</artifactId>
+</dependency>
+
+<dependency>
+    <groupId>com.yubico</groupId>
+    <artifactId>webauthn-server-core</artifactId>
+    <version>2.5.0</version>
+</dependency>
+
+<dependency>
+    <groupId>org.projectlombok</groupId>
+    <artifactId>lombok</artifactId>
+    <scope>provided</scope>
+</dependency>
+
+<dependency>
+    <groupId>com.h2database</groupId>
+    <artifactId>h2</artifactId>
+</dependency>
+
+<dependency>
+    <groupId>org.flywaydb</groupId>
+    <artifactId>flyway-core</artifactId>
+</dependency>
+```
+
+### Database
+We use Flyway to set up our database; we need to keep track of our users and their passkeys.
+
+We need to save the following data for the user.
+
+The username is the identifier specified by the user, and the user handle is the unique identifier generated by the system, used to streamline communication between the FIDO server and the authenticator.\
+The assertion contains information such as user identification, the authentication method used, and cryptographic signatures to ensure the authenticity of the claim. It is used to confirm that the user has been successfully authenticated and that the authenticator is indeed trustworthy. It is also rebuilt every time a user attempts to log in.\
+The public key JSON is used when creating a new public key for a user, typically during the registration process. It contains all the parameters and options necessary for generating a new public key for the respective user. The authenticator will use these options to generate the new key.\
```

Comment:
Confusing -> rewrite.

Reply:
Explain what is generated where and when. This is unclear now (the authenticator = server? doesn't generate a public key per user right? The "user"/"user-device" does that.